release-22.2: logictestccl: Fix regional_by_row_query_behavior flake

msirek opened the pull request:
Backport:
  * 1/1 commits from "logictestccl: Fix regional_by_row_query_behavior flake" (#87391)
  * 1/1 commits from "logictest: check for nil SystemConfig before purging zone config cache" (#90802)
  * 1/1 commits from "config: acquire write lock when overwriting the map" (#92606)

Please see individual PRs for details.

/cc @cockroachdb/release

----
logictest: only purge zone configs for in-memory test clusters

This commit backports code needed to fix #104466.

Informs: #104466

Release note: None

----

logictest: retry support for "statement" and "query" commands
This adds a retry command to logic tests which may be issued on a
separate line preceding either a "statement" or "query" command.
It causes the statement to be retried with exponential backoff up
to some maximum duration, e.g.
```
retry
statement error column "non_exist" does not exist
ALTER TABLE created_as_global SET LOCALITY REGIONAL BY ROW AS "non_exist"
```
This has the same effect as the retry option of the query command, but
now also supports "statement ok", "statement error" and "query error"
commands.

Retry of a query command may be specified by the standalone retry
command, the retry option of the query command, or both.

Fixes: https://github.com/cockroachdb/cockroach/issues/95668

Epic: [CRDB-20535](https://cockroachlabs.atlassian.net/browse/CRDB-20535)

Release note: None

----
logictest: retry support for "statement" and "query" commands

This adds a retry command to logic tests which may be issued on a
separate line preceding either a "statement" or "query" command.
It causes the statement to be retried with exponential backoff up
to some maximum duration, e.g.
```
retry
statement error column "non_exist" does not exist
ALTER TABLE created_as_global SET LOCALITY REGIONAL BY ROW AS "non_exist"
```
This has the same effect as the retry option of the query command, but
now also supports "statement ok", "statement error" and "query error"
commands.

Retry of a query command may be specified by the standalone retry
command, the retry option of the query command, or both.

Fixes: #95668

Epic: CRDB-20535

Release note: None

----
logictestccl: add more retries to logic tests

This commit adds more retries to multiregion logic tests.

Fixes #104466

Release note: None

----
Release justification: Fix #104466

Commit:
logictestccl: Fix regional_by_row_query_behavior flake

Fixes #86284

This fixes a test flake in regional_by_row_query_behavior which is
caused by 2 issues:

  1. Zone configs not becoming available in `SystemConfig` between
     table creation time and query execution time.

  2. Metamorphic testing modify the `coldata-batch-size` setting,
     causing a different number of kv reads to be done, which affects
     the kv trace results in this test.

The flake can be reproduced via the command:
```
./dev testlogic ccl --files=regional_by_row_query_behavior \
                    --config=multiregion-9node-3region-3azs --stress
```
For point 1, the `retry` logic test directive is enhanced to purge the
zone config cache in the SystemConfig when a failing test is retried,
allowing the updated zone config to be read.

For point 2, the logic test blocking flag `!metamorphic` which resets
batch size related settings that were modified in metamorphic mode back
to their default production values, is updated to also cover the
`coldata-batch-size` setting. The flag is renamed to
`!metamorphic-batch-sizes` to better reflect its actual effect.

Release justification: low risk fix for test flake

Release note: None

Commit:
logictest: check for nil SystemConfig before purging zone config cache

Fixes #88398

This fixes a panic which can occur duing cluster startup in a logic test
or during retry of a test case in a logic test when an attempt is made
to purge the zone config cache and no SystemConfig is available.

Release note: None

yuzefovich committed:
config: acquire write lock when overwriting the map

Previously, by mistake we were acquiring a read lock even though we do
write operations in `PurgeZoneConfigCache`.

Release note: None

Commit:
logictest: retry support for "statement" and "query" commands

This adds a retry command to logic tests which may be issued on a
separate line preceding either a "statement" or "query" command.
It causes the statement to be retried with exponential backoff up
to some maximum duration, e.g.
```
retry
statement error column "non_exist" does not exist
ALTER TABLE created_as_global SET LOCALITY REGIONAL BY ROW AS "non_exist"
```
This has the same effect as the retry option of the query command, but
now also supports "statement ok", "statement error" and "query error"
commands.

Retry of a query command may be specified by the standalone retry
command, the retry option of the query command, or both.

Fixes: #95668

Epic: CRDB-20535

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105787)
<!-- Reviewable:end -->

Commit:
logictest: only purge zone configs for in-memory test clusters

This commit backports code needed to fix #104466.

Informs: #104466

Release note: None

Commit:
logictestccl: add more retries to logic tests

This commit adds more retries to multiregion logic tests.

Fixes #104466

Release note: None

yuzefovich reviewed (approved):
> Reviewed 16 of 16 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105787)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek)



<!-- Sent from Reviewable.io -->

msirek commented:
TFTR!
